pkg/metrics/loadbalancer.go

```diff
@@ -0,0 +1,82 @@
+// SPDX-License-Identifier: MIT
+
+package metrics
+
+import (
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/loadbalancer/v2/loadbalancers"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/klog"
+)
+
+var (
+	loadbalancerAdminStateUp *prometheus.GaugeVec
+	loadbalancerStatus       *prometheus.GaugeVec
+
+	// possible load balancer provisioning states, from https://github.com/openstack/octavia-lib/blob/fe022cdf14604206af783c8a0887c008c48fd053/octavia_lib/common/constants.py#L169
+	provisioningStates = []string{"ALLOCATED", "BOOTING", "READY", "ACTIVE", "PENDING_DELETE", "PENDING_UPDATE", "PENDING_CREATE", "DELETED", "ERROR"}
+
+	loadBalancerLabels = []string{"id", "vip_address", "provider", "port_id"}
+)
+
+func registerLoadBalancerMetrics() {
+	loadbalancerAdminStateUp = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: generateName("loadbalancer_admin_state_up"),
+			Help: "Load balancer status",
+		},
+		loadBalancerLabels,
+	)
+	loadbalancerStatus = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: generateName("loadbalancer_provisioning_status"),
+			Help: "Load balancer status",
+		},
+		append(loadBalancerLabels, "provisioning_status"),
+	)
+
+	prometheus.MustRegister(loadbalancerAdminStateUp)
+	prometheus.MustRegister(loadbalancerStatus)
+}
+
+// ScrapeLoadBalancerMetrics makes the list request to the load balancer api and
+// passes the result to a scrape function.
+func ScrapeLoadBalancerMetrics(client *gophercloud.ServiceClient, tenantID string) error {
+	// first step: gather the data
+	mc := newOpenStackMetric("loadbalancer", "list")
+	pages, err := loadbalancers.List(client, loadbalancers.ListOpts{}).AllPages()
+	if mc.Observe(err) != nil {
+		// only warn, maybe the next scrape will work.
+		klog.Warningf("Unable to scrape floating ips: %v", err)
+		return err
+	}
+	loadBalancerList, err := loadbalancers.ExtractLoadBalancers(pages)
+	if err != nil {
+		// only warn, maybe the next scrape will work.
+		klog.Warningf("Unable to scrape load balancers: %v", err)
+		return err
+	}
+
+	// second step: reset the old metrics
+	loadbalancerStatus.Reset()
+
+	// third step: publish the metrics
+	for _, lb := range loadBalancerList {
+		scrapeLoadBalancerMetric(lb)
+	}
+
+	return nil
+}
+
+// scrapeLoadBalancerMetric extracts data from a floating ip and exposes the metrics via prometheus
+func scrapeLoadBalancerMetric(lb loadbalancers.LoadBalancer) {
+	labels := []string{lb.ID, lb.VipAddress, lb.Provider, lb.VipPortID}
+
+	loadbalancerAdminStateUp.WithLabelValues(labels...).Set(boolFloat64(lb.AdminStateUp))
+
+	// create one metric per provisioning status
+	for _, state := range provisioningStates {
+		stateLabels := append(labels, state)
+		loadbalancerStatus.WithLabelValues(stateLabels...).Set(boolFloat64(lb.ProvisioningStatus == state))
```

Comment:
Why do we only want to know the hard coded possible provisioning states? I think it's okay to add all states as a metric.

Reply:
because this are all provisioning states which could occur.

We will have one metric per state afterwards, it's the same way we do for cinder volume states. Example from the example metrics:

```
kos_loadbalancer_provisioning_status{id="059e7a29-5229-409e-870d-6ceb9c1059a9",port_id="10902a71-299d-4666-b086-1a0725288dac",provider="vmwareedge",provisioning_status="ACTIVE",vip_address="10.6.0.8"} 1
kos_loadbalancer_provisioning_status{id="059e7a29-5229-409e-870d-6ceb9c1059a9",port_id="10902a71-299d-4666-b086-1a0725288dac",provider="vmwareedge",provisioning_status="ALLOCATED",vip_address="10.6.0.8"} 0
kos_loadbalancer_provisioning_status{id="059e7a29-5229-409e-870d-6ceb9c1059a9",port_id="10902a71-299d-4666-b086-1a0725288dac",provider="vmwareedge",provisioning_status="BOOTING",vip_address="10.6.0.8"} 0
kos_loadbalancer_provisioning_status{id="059e7a29-5229-409e-870d-6ceb9c1059a9",port_id="10902a71-299d-4666-b086-1a0725288dac",provider="vmwareedge",provisioning_status="DELETED",vip_address="10.6.0.8"} 0
kos_loadbalancer_provisioning_status{id="059e7a29-5229-409e-870d-6ceb9c1059a9",port_id="10902a71-299d-4666-b086-1a0725288dac",provider="vmwareedge",provisioning_status="ERROR",vip_address="10.6.0.8"} 0
kos_loadbalancer_provisioning_status{id="059e7a29-5229-409e-870d-6ceb9c1059a9",port_id="10902a71-299d-4666-b086-1a0725288dac",provider="vmwareedge",provisioning_status="PENDING_CREATE",vip_address="10.6.0.8"} 0
kos_loadbalancer_provisioning_status{id="059e7a29-5229-409e-870d-6ceb9c1059a9",port_id="10902a71-299d-4666-b086-1a0725288dac",provider="vmwareedge",provisioning_status="PENDING_DELETE",vip_address="10.6.0.8"} 0
kos_loadbalancer_provisioning_status{id="059e7a29-5229-409e-870d-6ceb9c1059a9",port_id="10902a71-299d-4666-b086-1a0725288dac",provider="vmwareedge",provisioning_status="PENDING_UPDATE",vip_address="10.6.0.8"} 0
kos_loadbalancer_provisioning_status{id="059e7a29-5229-409e-870d-6ceb9c1059a9",port_id="10902a71-299d-4666-b086-1a0725288dac",provider="vmwareedge",provisioning_status="READY",vip_address="10.6.0.8"} 0
```

Reply:
Yes, that's true. But what if some day additional states will be available? I mean, I just don't think it's necessary to filter here. There is no benefit imho, but okay for me

Reply:
Normally: the api won't change here, also if we don't filter we will have a single metric and cannot add some alerts matching a single state because the metric would just be gone and not say that it is not in this state anymore.

Reply:
Ahh, now I get it. Okay, good point :+1:

Reply:
I just thought about the `lb.ProvisioningStatus == state`, not the entire loop